### ISSUE TITLE
v0.12 testing: Memory consumption appears to be significantly higher after moving to v0.12 plugin SDKs

### ISSUE THREAD

**Issue by alexng-canuck** [hashicorp/terraform#21104]
### Terraform Version

```
v0.12.0-alpha4.0.20190417210818-177a7afb781f
```

### Terraform Configuration Files

```hcl
Can be shared if it's relevant.
```

### Debug Output

Goland Memory Profiles can be shared upon request.

### Expected Behavior

With v0.11 SDK, our test case allocates about ~40 MB according to Goland memory profiling.

### Actual Behavior

With v0.12 SDK, the same test allocates about ~300MB.

### Additional Context

After moving to v0.12 SDK, we had tests encounter an OOM panic when running inside docker containers. These tests were passing when still using v0.11 SDK.

**Comment by alexng-canuck** [comment]
With v0.12
<img width="1331" alt="Screen Shot 2019-04-24 at 5 55 34 PM" src="https://user-images.githubusercontent.com/9562037/56702716-53f49980-66ba-11e9-9bf5-50f7cc10f0b6.png">

With v0.11
<img width="1265" alt="Screen Shot 2019-04-24 at 5 56 18 PM" src="https://user-images.githubusercontent.com/9562037/56702717-53f49980-66ba-11e9-9dde-9611408a2ff3.png">

**Comment by alexng-canuck** [comment]
[MemoryProfiles.zip](https://github.com/hashicorp/terraform/files/3114683/MemoryProfiles.zip)

**Comment by paultyng** [comment]
Have you tried this outside the test framework using the beta2 CLI and a provider binary? I'm curious what part of this is provider side and what part of this is core side, as the test framework includes both, it probably artificially inflates this amount.

**Comment by alexng-canuck** [comment]
@paultyng 
Update: 
I suspect the issue is related to not cleaning up the test GRPC servers that are started by the `GRPCTestProvider` shim and the memory leaks pile up when a lot of test cases are run from a single `go test` command.

I did a little experimental hack where I keep a reference to the test GRPC server and call a `GracefulStop()` on it whenever `GRPCProvider.Close()` is invoked by Terraform's close provider evaluation.

Without this change, my `go test` command grows to about 6 GB at peak usage. After this change, my `go test` command's peak memory usage is reduced to about 700 MB (which was more in line with the v0.11 plugin SDK).

I'm encountering locking/wait issues with this hack however, which is probably beyond my ability to investigate. Would be great if the memory leak can be fixed in a more graceful manner before v0.12 GA. Otherwise, I'm not sure how CI testing doesn't hit OOM panics, since I believe tests are all invoked from the same `go test` command.
